README.md
<a href="https://github.com/ballandt/avmath/wiki">
  <img src="docs/avmath.svg" width="400" title="avmath">
</a>

---
![GitHub branch checks state](https://img.shields.io/github/checks-status/ballandt/avmath/master)
![GitHub release (latest by date including pre-releases)](https://img.shields.io/github/v/release/ballandt/avmath?include_prereleases)
![PyPI](https://img.shields.io/pypi/v/avmath)
![PyPI - Downloads](https://img.shields.io/pypi/dm/avmath)
![Libraries.io SourceRank](https://img.shields.io/librariesio/sourcerank/pypi/avmath)

Avmath is a library project that uses python algorithms to numerically solve
math problems. Its interface is based on mathematical habits of
writing. It particularly concentrates on the concepts of linear algebra
and analysis, but also generates its own functions based on Taylor-series.
For further description, see the [wiki](https://github.com/ballandt/avmath/wiki).

---
## Contents

* [Description](https://github.com/ballandt/avmath/wiki)
  * [Features](https://github.com/ballandt/avmath/wiki#features)
  * [Developments](https://github.com/ballandt/avmath/wiki/developments)
* [Documentation](https://github.com/ballandt/avmath/wiki/Documentation)
  * [Installation](https://github.com/ballandt/avmath/wiki/Documentation#installation)
* [Contribute](https://github.com/ballandt/avmath/blob/master/CONTRIBUTING.md)
* [Changelog](https://github.com/ballandt/avmath/blob/master/CHANGELOG.md)
* [Releases](https://www.github.com/ballandt/avmath/releases)
---

## Information

| Category           | Data                                                                 |
|--------------------|----------------------------------------------------------------------|
| Name               | avmath                                                               |
| PyPi name          | [avmath](https://pypi.org/project/avmath/)                           |
| Author             | Camillo Ballandt                                                     |
| Release version    | [3.1.1](https://www.github.com/ballandt/avmath/releases/tag/v3.1.1)  |
| Developing version | [Developments](https://github.com/ballandt/avmath/wiki/Developments) |
| Requirements       | Python 3.10                                                          |


CHANGELOG.md
# AdVanced Math Changelog

## Contents

| Release                           | Name           |       Date | Integral developments                                           |
|:----------------------------------|----------------|-----------:|-----------------------------------------------------------------|
| [3.1.1](#311-2022-01-08)          | Ondes bugfix 1 | 2022-01-09 | fixes for derivatives                                           |
| [3.1.0](#310-2022-01-01)          | Ondes          | 2022-01-01 | Completions in `analysis`                                       |
| [3.0.0](#300-2021-12-01)          | Illumination   | 2021-12-01 | implementation of `Fraction`, Gauss-features for matrices       |
| [2.0.0](#200-2021-10-24)          | Repensé        | 2021-10-24 | independence of math module with Taylor-series, matrix features |
| [2.0.0-b2](#200-beta2-2021-09-19) | Repensé beta 2 | 2021-09-19 | not yet stable version of 2.0.0                                 |
| [1.2.0](#120-2021-09-10)          | Amitié         | 2021-09-10 | coordination of elements                                        |
| [1.1.0](#110-2021-09-01)          | First release  | 2021-09-01 | `Angle` class in former `geo`                                   |
| [1.0.0](#100-2021-09-01)          | Initial avmath | 2021-09-01 | Upload to GitHub                                                |

---
## 3.2.0 (2022-04-01)

Not yet published

### Minor changes

* `__init__`
  * `Fraction`
    * added own `__rtruediv__` method to fix ([#9][i9])
* `algebra`
  * `Vector`
    * new method `orthogonal`
      * transposes given vectors to orthogonal vectors of `self`
  * `Matrix`
    * new method `qr`
      * executes qr decomposition
    * new method `eigenvalues`
      * returns the eigenvalues of a matrix
    * new method `eigenvector`
      * returns the eigenvector to a given eigenvalue of a matrix
* `analysis`
  * new class `Polynomial`

### Patch level changes

* `__init__`
  * `Fraction`
    * `__truediv__`
      * fixed division by integer ([#1][i1])
    * `__mul__`
      * fixed condition branches ([#3][i3])
  * `gcd`
    * `gcd([x], 0)` returns x ([#7][i7])
  * `lcm`
    * `lcm([x], 0)` returns 0 ([#7][i7])
* `algebra`
  * `Matrix`
    * `append`
      * fixed column append ([#5][i5])

---
## 3.1.1 (2022-01-08)

[Find release](https://github.com/ballandt/avmath/releases/tag/v3.1.1)

#### Patch level changes

* PEP8 compatible line length 79
* `analysis`
  * `Function`
    * `derivative`
      * h calculation does not fail if f(x) = 0
* `__init__`
  * added `sgn` function to `scope`
  * removed `Fraction` scope because of errors in `Function.derivative` and `integral`
  * `sinh` and `cosh`:
    * x values with |x| > 710 cause `ArgumentError`
  * constants got typehints
    * `pi`
    * `e`
    * `phi`
    * `gamma`

## 3.1.0 (2022-01-01)

[Find release](https://github.com/ballandt/avmath/releases/tag/v3.1.0)

#### Announcement

* `analysis`
  * `Function`
    * The methods `num_dif`, `second_num_dif` and `num_int` are abandoned for `derivative`, `second_derivative` and `integral` yet are still usable
  * `Polynom` is cancelled

#### Minor changes

* `__init__`
  * `Fraction`
    * `__init__`
      * fractions can be used for initialisation
    * `complex` method to convert Fraction to complex
* `analysis`
  * `Function`
    * new method `newton_method_extrema`
      * uses newton method with differentiations for the calculation of extrema
    * new method `derivative`
      * new method for function derivatives
      * preferred over `numd_dif`
    * new method `second_derivative`
      * method for second derivative
      * preferred over `second_num_dif`
    * new method `integral`
      * calculates numerical integral
      * in default case uses rectangular rule
      * `option="trapeze"` instructs to use trapeze rule
      * preferred over `num_int`
    * new method `__rtruediv__` for reversed division
    * new method `tangent`
      * constructs a tangent on a function at a given x
      * returns a linear `Function` (y = a * x + b)
    * new method `normal`
      * constructs a normal on a function at a given x
      * returns a linear `Function` (y = a * x + b)

#### Patch level changes

* `__init__`
  * `Fraction`
    * `__truediv__`
      * formula fixes
    * `__add__` supports complex
    * `__sub__` supports complex
  * `arcsin`
    * exception fixed for values < -1
* `analysis`
  * `Function`
    * `max`-method uses newton method and is much faster and more accurate
    * `numdif` developments for accuracy
    * `replace`-method does not crash while handling x with no parameter
    * `__mul__` can handle REAL argument
    * `__truediv__` can handle REAL argument
    * `__rtruediv__` can handle REAL argument
    * `newton_method` takes a `1e-2` step if the derivative is 0
* `algebra`
  * `Structure`
    * `area`
      * fixes for angles > 180°

---
## 3.0.0 (2021-12-01)

[Find release](https://github.com/ballandt/avmath/releases/tag/v3.0.0)

#### API Changes

* `__init__`
  * redesigned function `avmath.pi()` to constant `avmath.pi`
  * redesigned function `avmath.e()` to constant `avmath.e`
  * `_FLOAT_EQ`
    * __Deletion__
  * `_Point`
    * __Deletion__
  * `DimensionError`
    * moved from `algebra`
* `algebra`
  * `Angle`
    * __Deletion__
  * `Vector`
    * `avmath.algebra.Vector.angle()` returns float
    * changed location of vector product from `__pow__` to `cross`
    * `pow` returns vector multiplied n times by itself
  * `Straight`
    * __Deletion__
  * `Point` -> `Tuple`
    * renamed `Point` to `Tuple`
  * `AmountOfDimensionsError`
    * __Deletion__
    * `DimensionError`gets called instead.
  * `MatrixMultiplicationError`
    * __Deletion__
    * `MatrixError` gets called instead
  * `Matrix`
    * `row`
      * returns `Vector` type
    * `column`
      * returns `Vector` type
  * `Structure`
    * `circ` -> `circumference`
      * renamed `circ` to `circumference`
* `analysis`

src/avmath/__init__.py
"""AVMATH

Avmath is a Python module for mathematical purposes.
It contains functionalities for simple mathematical
usage as sine or logarithm functions and submodules
for more advanced applied math in the topics of ana-
lysis and linear algebra.

The module uses a mathematical syntax. The classes
and functions are named in a mathematical manner and
the class operations enable a mathematical workflow.
Its issue is to calculate as accurate as possible and
not firstly the aspect of time.

Documentation: https://github.com/ballandt/avmath/wiki/documentation
GitHub: https://www.github.com/ballandt/avmath
PyPi: https://www.pypi.org/project/avmath
"""

__author__ = "Camillo Ballandt"
__version__ = "3.1.1"
__date__ = "2022/01/08"

__all__ = ["Fraction",
           "sin", "cos", "tan",
           "arcsin", "arccos", "arctan",
           "sinh", "cosh", "tanh",
           "arsinh", "arcosh", "artanh",
           "ln", "log", "is_even", "fac", "sgn",
           "pi", "e", "phi", "gamma"]

import time
from typing import Union as _Union, Iterable as _Iterable

_TAYLOR_DIFFERENCE = 1e-16
_MAX_CALCULATION_TIME = 5
REAL = _Union[int, float, 'Fraction']

e: float = 2.718_281_828_459_045_235_360
pi: float = 3.141_592_653_589_793_238_463
phi: float = 1.618_033_988_749_894_848_205
gamma: float = 0.577_215_664_901_532_860_607

two_digit_primes: list[int] = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41,
                               43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]


class ArgumentError(Exception):
    """Raised if false argument is given."""

    def __init__(self, got, want):
        self.got = got
        self.want = want

    def __str__(self):
        return f"False argument given. Expected {self.want}, got {self.got}."


class DimensionError(Exception):
    """Raised if arguments have different amount of dimensions."""

    def __init__(self, got=None, want=None, other=None):
        if other is None:
            self.got = got
            self.want = want
        else:
            self.other = other

    def __str__(self):
        try:
            return f"Wrong amount of dimensions." \
                   f"Expected {self.want}; got {self.got}"
        except AttributeError:
            return self.other


class Fraction:
    """Mathematical fraction"""

    def __init__(self, numerator: REAL, denominator: REAL):
        """Initializes fraction. Denominator will be made positive.
        Insert
        Fraction(a, b)
        for
        a/b

        and
        Fraction(a, Fraction(b, c))
        for
        a/(b/c)
        """
        if denominator == 0:
            raise ZeroDivisionError("Denominator must not be zero")
        if type(numerator) == int and type(denominator) == int:
            self.a = numerator * sgn(denominator)
            self.b = abs(denominator)
        elif type(numerator) == Fraction or type(denominator) == Fraction:
            self.a = (numerator / denominator).a
            self.b = (numerator / denominator).b
        elif type(numerator) == float or type(denominator) == float:
            numerator, denominator = float(numerator), float(denominator)
            numerators = numerator.as_integer_ratio()
            denominators = denominator.as_integer_ratio()
            self.a = numerators[0] * denominators[1]
            self.b = numerators[1] * denominators[0]

    def __repr__(self) -> str:
        """Returns string representation. Always reduced."""
        if self.a / self.b == self.a // self.b:
            ret_str = f"{self.a // self.b}"
        else:
            ret_str = f"{self.reduce().a}/{self.reduce().b}"
        return ret_str

    def __neg__(self) -> 'Fraction':
        """Returns negative fraction."""
        return Fraction(-self.a, self.b)

    def __eq__(self, other: REAL) -> bool:
        """Verifies the equality of two fractions."""
        return self.reduce().a == other.reduce().a \
            and self.reduce().b == other.reduce().b

    def __lt__(self, other: REAL) -> bool:
        """Less than."""
        return float(self) < float(other)

    def __gt__(self, other: REAL) -> bool:
        """Greater than."""
        return float(other) < float(self)

    def __add__(self, other: REAL | complex) -> 'Fraction':
        """Adds either two fractions or fractions and numbers."""
        if type(other) in (int, float):
            a = self.a + other * self.b
            return Fraction(a, self.b)

        elif type(other) == complex:
            return complex(self) + other

        elif type(other) == Fraction:
            reduced_self = self.reduce()
            reduced_other = other.reduce()
            factor = lcm(reduced_self.b, reduced_other.b)
            summand1 = Fraction(reduced_self.a*int(factor/reduced_self.b),
                                reduced_self.b*int(factor/reduced_self.b))
            summand2 = Fraction(reduced_other.a*int(factor/reduced_other.b),
                                reduced_other.b*int(factor/reduced_other.b))
            return Fraction(summand1.a + summand2.a, summand1.b)

    __radd__ = __add__

    def __sub__(self, other: REAL) -> 'Fraction':
        """Subtracts either real from Fraction or Fraction from Fraction."""
        return self + -other

    def __rsub__(self, other: REAL) -> 'Fraction':
        return other + -self

    def __mul__(self, other: REAL) -> 'Fraction':
        """Multiplies REALS."""
        if type(other) in (int, float):
            return Fraction(self.a * other, self.b)
        elif type(other) == Fraction:
            return Fraction(self.a * other.a, self.b * other.b).reduce()
        else:
            return other.__mul__(self)

    __rmul__ = __mul__

    def __truediv__(self, other: REAL) -> 'Fraction':
        if type(other) == Fraction:
            res = self * other ** -1
        else:
            res = Fraction(self.a, self.b * other)
        if type(res) == Fraction:
            res = res.reduce()
        return res

    def __rtruediv__(self, other):
        res = Fraction(self.b * other, self.a)
        if type(res) == Fraction:
            res = res.reduce()
        return res

    def __pow__(self, power: REAL) -> 'Fraction':
        if power == -1:
            return Fraction(self.b, self.a)
        else:
            return Fraction(self.a ** float(power), self.b ** float(power))

    def __rpow__(self, other: REAL) -> float:
        return (other ** self.a) ** (1 / self.b)

    def __mod__(self, other: REAL) -> float:
        return float(self) % float(other)

    __rmod__ = __mod__

    def __int__(self) -> int:
        return self.a // self.b

    def __float__(self) -> float:
        reduced_fraction = self.reduce()
        return reduced_fraction.a / reduced_fraction.b

    def __complex__(self):
        return float(self) + 0j

    def __abs__(self) -> 'Fraction':
        return Fraction(abs(self.a), self.b)

    def reduce(self) -> 'Fraction':
        divisor = gcd(self.a, self.b)
        return Fraction(int(self.a / divisor), int(self.b / divisor))

    def int_args(self) -> bool:
        return type(self.a) == int and type(self.b) == int


def _check_types(arg: _Iterable, *types):
    """Checks if the elements of the argument belong to the given types."""
    for ele in arg:
        if not type(ele) in types:
            raise ArgumentError(type(ele), types)
    return True


def is_even(x: int) -> bool:
    """Checks if x is an even number"""
    return x/2 == x // 2


def is_prime(x: int) -> bool:
    """Checks if integer is prime."""
    int_root = int(x ** 0.5)
    for i in range(2, int_root):
        if x / i == x // i:
            return False
    return True


def gcd(x: int, y: int) -> int:
    """Greatest common divisor."""
    if y == 0:
        return x
    while x % y != 0:
        r = x % y
        x = y
        y = r
    return abs(y)


def lcm(x: int, y: int) -> int:
    """Least common multiply."""
    return int(abs(x * y) / gcd(x, y)) if y != 0 else 0


def sgn(x: REAL) -> int:
    """Returns signum of x."""
    if x < 0:
        return -1
    elif x == 0:
        return 0
    elif x > 0:
        return 1


def fac(x: REAL, opt: str = None):
    """Returns faculty of x.
    fac(x)               is x!
    fac(x, opt="double") is x!!
    """
    if x < 0:
        raise ArgumentError("x < 0", "x >= 0")
    if int(x) != x:
        raise ArgumentError("real x", "natural x")
    x = int(x)
    res = 1
    if opt == "double":
        if not is_even(x):
            for i in range(int(x / 2) + 1):
                res *= 2 * i + 1
        else:
            for i in range(1, int(x / 2)+1):
                res *= 2*i
    else:
        for i in range(1, x + 1):
            res *= i
    return res


def ln(x: REAL) -> float:
    """Natural logarithm."""
    if x <= 0:
        raise ArgumentError(x, "x >= 0")
    summand = 0
    while x > e:
        x /= e
        summand += 1
    while x < 1/e:
        x *= e
        summand -= 1
    res = 0
    k = 0
    while True:
        mem_res = res
        res += (x - 1) ** (2 * k + 1) / ((2 * k + 1) * (x + 1) ** (2 * k + 1))
        if abs(mem_res - res) < _TAYLOR_DIFFERENCE:
            break
        k += 1
    return 2 * res + summand


def log(x: REAL, base: REAL) -> float:
    """Logarithm."""
    return ln(x) / ln(base)


def sin(x: REAL) -> float:
    """Sine."""
    x %= 2 * pi
    res = 0
    k = 0
    while True:
        mem_res = res
        res += (-1) ** k * x ** (2 * k + 1) / fac(2 * k + 1)
        if abs(mem_res - res) < _TAYLOR_DIFFERENCE:
            return res
        k += 1


def cos(x: REAL) -> float:
    """Cosine."""
    x %= 2 * pi
    res = 0
    k = 0
    while True:
        mem_res = res
        res += (-1) ** k * (x**(2 * k) / fac(2 * k))
        if abs(mem_res - res) < _TAYLOR_DIFFERENCE:
            return res
        k += 1


def tan(x: REAL) -> float:
    """Tangent."""
    return sin(x) / cos(x)


def arcsin(x: REAL) -> float:
    """Arc sine."""
    if abs(x) > 1:
        raise ArgumentError("x > 1", "x <= 1")
    if abs(x) == 1:
        return sgn(x) * pi / 2
    res = x
    k = 1
    start_time = time.time()
    while (time.time() - start_time) < _MAX_CALCULATION_TIME:
        mem_res = res
        res += fac(2 * k - 1, opt="double") /\
            (fac(2 * k, opt="double") * (2 * k + 1)) * x ** (2 * k + 1)
        if abs(mem_res - res) < _TAYLOR_DIFFERENCE:
            break
        k += 1
    return res


def arccos(x: REAL) -> float:
    """Arc cosine."""
    return pi/2 - arcsin(x)


def arctan(x: REAL) -> float:
    """Arc tangent."""
    res = 0
    k = 0
    if abs(x) <= 1:
        start_time = time.time()
        while (time.time() - start_time) < _MAX_CALCULATION_TIME:
            mem_res = res
            res += (-1)**k * x**(2*k + 1) / (2*k + 1)
            if abs(mem_res - res) < _TAYLOR_DIFFERENCE:
                break
            k += 1
    else:
        if x >= 0:
            res = pi / 2
        else:
            res = -pi / 2
        start_time = time.time()
        while (time.time() - start_time) < _MAX_CALCULATION_TIME:
            mem_res = res
            res += (-1)**(k + 1) / ((2*k + 1) * x**(2*k + 1))
            if abs(mem_res - res) < _TAYLOR_DIFFERENCE:
                break
            k += 1
    return res


def sinh(x: REAL) -> float:
    """Hyperbolic sine."""
    if abs(x) > 710:
        raise ArgumentError(x, "argument |x| < 710")
    res = 0
    k = 0
    while True:
        mem_res = res
        res += x ** (2 * k + 1) / fac(2 * k + 1)
        if abs(mem_res - res) < _TAYLOR_DIFFERENCE:
            return res
        k += 1


def cosh(x: REAL) -> float:
    """Hyperbolic cosine."""
    if abs(x) > 710:
        raise ArgumentError(x, "argument |x| < 710")
    res = 0
    k = 0
    while True:
        mem_res = res
        res += x**(2*k) / fac(2*k)
        if abs(mem_res - res) < _TAYLOR_DIFFERENCE:
            return res
        k += 1


def tanh(x: REAL) -> float:
    """Hyperbolic tangent."""
    return sinh(x) / cosh(x) if abs(x) <= 20 else sgn(x) * 1.0


def arsinh(x: REAL) -> float:
    """Inverse hyperbolic sine."""
    return sgn(x) * ln(abs(x) + (x**2 + 1)**0.5)


def arcosh(x: REAL) -> float:
    """Inverse hyperbolic cosine."""
    if x < 1:
        raise ArgumentError(x, "x >= 1")
    return ln(x + (x**2 - 1)**0.5)


def artanh(x: REAL) -> float:
    """Inverse hyperbolic tangent."""
    if abs(x) >= 1:
        raise ArgumentError(x, "|x| < 1")
    return 0.5 * ln((1 + x) / (1 - x))


scope = {
    "sin": sin, "arcsin": arcsin,
    "cos": cos, "arccos": arccos,
    "tan": tan, "arctan": arctan,
    "sinh": sinh, "arsinh": arsinh,
    "cosh": cosh, "arccosh": arcosh,
    "tanh": tanh, "artanh": artanh,
    "log": log,
    "ln": ln,
    "fac": fac,
    "sgn": sgn,
    "e": e,
    "pi": pi,
}
